otwcode/otwarchive
# AO3-5565 Prevent users from bookmarking work multiple times

# Pull Request Checklist


## Issue

https://otwarchive.atlassian.net/browse/AO3-5565

## Purpose

~Code prevents users from bookmarking the same work for one pseud more than once.~
I am not sure that's the right approach though.

From the code design I assumed that different pseuds of the same user should still be able to bookmark same work.
However it is possible that bookmarks were designed to be connected to pseuds rather than to users for other reasons.

`index_bookmarkable_pseud` is also not made unique, I am unsure if it has to be, but rubocop throws `Rails/UniqueValidationWithoutIndex: Uniqueness validation should be with a unique index.`

Please let me know if this is wrong and the constraint somehow should be placed on user instead of the pseud, or if something must be done with the index.

-----
**Update**: 
Code prevents users from bookmarking the same work twice, regardless of used pseud. 
Database Index is not made unique.

## Testing Instructions

How can the Archive's QA team verify that this is working as you intended?

Before applying the code do:
1. Go to [root]/works/[id]/bookmarks/new
2. Create bookmark with pseudA - it should be created successuflly
3. Go to [root]/works/[id]/bookmarks/new again
4. Create second bookmark with pseudA or pseudB - it would be created successfully

After applying the code changes do:
1. Go to [root]/works/[id]/bookmarks/new
2. Create bookmark with pseudA or pseudB - error message should be visible
3. Update bookmarks created in the "Before" step - user should be able to do so, as well as reassign bookmarks from pseudB to pseudA and back.

## Credit

korrien, she/her
